sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java

```diff
@@ -786,6 +788,28 @@ public Node visitPrimaryKeyConstraint(SqlBaseParser.PrimaryKeyConstraintContext
         return new PrimaryKeyConstraint<>(visitCollection(context.columns().primaryExpression(), Expression.class));
     }
 
+    private Node visitCheckConstraint(SqlBaseParser.CheckConstraintContext ctx, String columnName) {
+        String name = null;
+        if (null != ctx.CONSTRAINT()) {
+            name = getIdentText(ctx.name);
+        }
+        Expression expression = (Expression) visit(ctx.expression);
+        return null == columnName ?
+            new CheckConstraint<>(name, expression) : new CheckColumnConstraint<>(name, columnName, expression);
+    }
+
+    @Override
+    public Node visitTableCheckConstraint(SqlBaseParser.TableCheckConstraintContext context) {
+        return visitCheckConstraint(context.checkConstraint(), null);
+    }
+
+    @Override
+    public Node visitColumnCheckConstraint(SqlBaseParser.ColumnCheckConstraintContext context) {
+        SqlBaseParser.ColumnDefinitionContext columnCtx = (SqlBaseParser.ColumnDefinitionContext) context.parent;
```

Comment:
was wondering whether it is absolutely necessary to access the parent node context from the child node? imho, it would be nice to keep the scope of node visitor method limited to the node context.

Reply:
It can be done, in fact we do not really need to have a column name, as it is only used to generate a name when the user does not provide one explicitly. But I will do that when I am closer to the end.

Also, if Node was an interface (while somehow still enforcing the hashcode, equals and tostring override in “subclasses”) I could only use one class for both cases (column/table level definitions) by having this class implement TableElement and ColumnConstraint, which would also be interfaces.

Reply:
@kovrus I need to have two classes because users can define CHECK constraints at both the column and the table level:

```
CREATE TABLE t (
   id int PRIMARY KEY,
   name string CHECK(name != 'peter'),
   size int,
   CHECK(size >= 0)
)
```
Each class is under different hierarchies (ColumnConstraint, TableElement). This could be avoided by having these hierarchies be interfaces. As I am already in this situation, I gather the column name (this is the only way I can think of), and then I use the column name to produce the check-constraint's name. Later on, when you do `SHOW CREATE TABLE t` you will see that the constraints are defined at the table level. Were it not for the column name in the check-constaint's name, we would loose information, i.e. the fact that the original statement defined the constraint at the column level.

If it is ok to drop this knowledge, then I will proceed to downsize. But, my preference would be not to drop knowledge.

Reply:
maybe we can enrich (map) the constraint with a column name in the analyzer? I am just sceptical about traversing the AST within the AstBuilder traversing logic :) Imho, it makes parser grammar and node visitors logic harder to change and reason about.

Reply:
Followed your suggestion, thank you!